repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -117,38 +125,48 @@ class Session(
       return
     }
 
-    if (_statements(statementId).state.get() == StatementState.Available ||
-      _statements(statementId).state.get() == StatementState.Cancelled ||
-      _statements(statementId).state.get() == StatementState.Cancelling) {
+    if (_statements(statementId).state.get().isOneOf(
+      StatementState.Available, StatementState.Cancelled, StatementState.Cancelling)) {
       return
     } else {
       // statement 1 is running and statement 2 is waiting. User cancels
       // statement 2 then cancels statement 1. The 2nd cancel call will loop and block the 1st
       // cancel call since cancelExecutor is single threaded. To avoid this, set the statement
       // state to cancelled when cancelling a waiting statement.
-      _statements(statementId).state.compareAndSet(StatementState.Waiting, StatementState.Cancelled)
-      _statements(statementId).state.compareAndSet(
+      _statements(statementId).compareAndTransition(
+        StatementState.Waiting, StatementState.Cancelled)
+      _statements(statementId).compareAndTransition(
         StatementState.Running, StatementState.Cancelling)
     }
 
     info(s"Cancelling statement $statementId...")
 
-      Future {
-        val deadline = livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TIMEOUT).millis.fromNow
-        while (_statements(statementId).state.get() == StatementState.Cancelling) {
+    Future {
+      val deadline = livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TIMEOUT).millis.fromNow
+      val runnable = new Runnable {
+        override def run(): Unit = {
           if (deadline.isOverdue()) {
             info(s"Failed to cancel statement $statementId.")
-            _statements(statementId).state.compareAndSet(
+            _statements(statementId).compareAndTransition(
               StatementState.Cancelling, StatementState.Cancelled)
           } else {
             _sc.foreach(_.cancelJobGroup(statementId.toString))
           }
-          Thread.sleep(livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TRIGGER_INTERVAL))
-        }
-        if (_statements(statementId).state.get() == StatementState.Cancelled) {
-          info(s"Statement $statementId cancelled.")
+
+          try {
+            Thread.sleep(livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TRIGGER_INTERVAL))
+          } catch {
+            case _: InterruptedException => // Ignore interrupted exception.
+            case NonFatal(e) => throw e
+          }
         }
-      }(cancelExecutor)
+      }
+
+      while(_statements(statementId).checkStateAndExecute(StatementState.Cancelling, runnable)) { }
```

Comment:
This `while` loop doesn't change the semantics compared to original code. It is because we introduce bunch of state transition code, so it is updated to keep align.